utils/sipdigest.1
.\" sipdigest - compute digest for user's secret in a specific sip realm
.\" Copyright (c) 2009 David Sugar <dyfet@gnutelephony.org>
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc.,59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH sipdigest "1" "December 2009" "GNU SIP Witch" "GNU Telephony"
.SH NAME
sipdigest \- compute digest for user's secret in specific sip realm
.SH SYNOPSIS
.B sipdigest [-realm="string"] [-md5|-sha1|-rmd160] \fIUSER\fR \fISECRET\fR 
.br
.SH DESCRIPTION
This is a tool to compute secret digest values for a \fIUSER\fR account which 
has a \fISECRET\fR (password) so that the digest can then be safely placed in
the sipwitch XML config files so that the password is not exposed in a 
potentially readable file.  SIP digests are computed using the authentication 
realm of the sip server for a given \fIUSER\fR id and \fISECRET\fR (password).  
Hence, it is desirable to set the realm once, because changing it will also 
require changing all the user hashes.
.PP
By default, sipdigest uses the current sip authentication realm and algorithm
as set by the siprealm command, though alternate realms and hashing algorithms
may also be specified.  Since sipdigest can compute an arbitrary digest it
can be useful to build database entries or config files where secret hashes
must be stored. 
.SH OPTIONS
.TP
.B -md5
Select md5 digest algorithm.
.TP
.B -realm=STRING
Compute digest based on the
.I STRING
realm passed instead of current server's realm.
.TP
.B -rmd160
Select rmd160 digest algorithm.
.TP
.B -sha1
Select sha1 digest algorithm.
.TP
.B -rmd160
.I 
.SH "EXIT STATUS"
Any error in argument format will return an exit status of 3.  The command
will normally return with exit status of 0 after printing the requested
digest.
.SH AUTHOR
.B sipdigest
was written by David Sugar <dyfet@gnutelephony.org>.
.SH "REPORTING BUGS"
Report bugs to sipwitch-devel@gnu.org.
.SH COPYRIGHT
Copyright \(co 2009 David Sugar, Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.



NEWS
SIP Witch NEWS -- history of visible user changes. Updated 2008-04-15

SIP Witch is an official package of the GNU Project as of August 10th 2007.
GNU SIP Witch is also part of GNU Telephony & the GNU Telecom subsystem.

sipwitch 0.1 codebase initial public release with support of SIP registration,
multi-target registration, authentication peering over SIP, and registration
querying of user agents, basic call processing between registered user agents
including call distribution for multi-target registrations, and basic SIP
instant messaging support.

sipwitch 0.4 represents the first minimally deployable baseline sipwitch
server.  This release includes multipath calling, and basic call forwarding.
Also important with this release is stablization of core management api's,
including SWIG interfaces for local scripting languages, and xmlrpc cgi.
Basic management and administration services can now be built for sipwitch.
Support for changing server state is also complete, such as for "day"/"night"
modes, holidays, etc.  Basic server call statistics and call detail record
processing is completed.

sipwitch 0.5 introduces the concept of internodal calling and allows sipwitch
to transform itself into a sip redirection server for certain local call
destinations.  Internodal calling differs from dialing to remote uri's in that
the authentication record is pre-stuffed into the remote invite, hence, all
nodes must share the same sip realm.  This allows one sipwitch server to call a
"local" extension number that actually appears on a remote switch rather than
the one the user agent is registered to. Internodal references are generated
using pattern rules through the static <routing> table <redirect> entries.  The
forward secure calling domain plugin also makes use of internodal calling to
connect with the insecure domain.
  


utils/sipdigest.cpp
// Copyright (C) 2008 David Sugar, Tycho Softworks.
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

#include <ucommon/ucommon.h>
#include <sipwitch/digest.h>
#include <config.h>

using namespace UCOMMON_NAMESPACE;
using namespace SIPWITCH_NAMESPACE;
	
extern "C" int main(int argc, char **argv)
{
	char *realm = NULL;
	const char *user, *secret;
	const char *mode = "md5";
	char buffer[128];
	string_t digestbuf;

	while(NULL != *(++argv)) {
		if(!strcmp(*argv, "--")) {
			++argv;
			break;
		}

		if(!strncmp(*argv, "--", 2))
			++*argv;

		if(!strcmp(*argv, "-version")) {
			printf("sipdigest 0.1\n"
				"Copyright (C) 2008 David Sugar, Tycho Softworks\n"
				"License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n"
				"This is free software: you are free to change and redistribute it.\n"
				"There is NO WARRANTY, to the extent permitted by law.\n");
			exit(0);
		}

		if(!strcmp(*argv, "-r") || !strcmp(*argv, "-realm")) {
			if(NULL == *(++argv)) {
				fprintf(stderr, "*** sipdigest: realm option missing\n");
				exit(3);
			}
			realm = strdup(*argv);
			continue;
		}

		if(String::equal(*argv, "-md5") || String::equal(*argv, "-sha1") || String::equal(*argv, "-rmd160"))
		{
			mode = strdup(*argv + 1);
			continue;
		}

		if(!strncmp(*argv, "-realm=", 7)) {
			realm = strdup(*argv + 7);
			continue;
		}

		if(!strcmp(*argv, "-?") || !strcmp(*argv, "-h") || !strcmp(*argv, "-help")) {
usage:
			fprintf(stderr, "usage: sipdigest [options] userid secret\n"
				"[options]\n"
				"  -md5                 specify md5 computation\n"
				"  -realm \"string\"    specify realm for digest\n"
				"  -rmd160              specify rmd160 computation\n"
				"  -sha1                specify sha1 computation\n"
				"Report bugs to sipwitch-devel@gnu.org\n");
			exit(3);
		}

		if(**argv == '-') {
			fprintf(stderr, "*** sipdigest: %s: unknown option\n", *argv);
			exit(3);
		}

		break;
	}

	user = *(argv++);
	if(!user) {
		fprintf(stderr, "*** sipdigest: userid missing\n");
		exit(3);
	}

	secret = *(argv++);
	if(!secret) {
		fprintf(stderr, "*** sipdigest: secret missing\n");
		exit(3);
	}

	if(!realm) {
		fsys_t fs;
		fsys::open(fs, DEFAULT_CFGPATH "/siprealm", fsys::ACCESS_RDONLY);
		if(!is(fs))
			fsys::open(fs, DEFAULT_VARPATH "/lib/sipwitch/uuid", fsys::ACCESS_RDONLY);

		if(!is(fs)) {
			fprintf(stderr, "*** sipdigest: no public realm known\n");
			exit(4);
		}
		memset(buffer, 0, sizeof(buffer));
		fsys::read(fs, buffer, sizeof(buffer) - 1);
		fsys::close(fs);

		char *cp = strchr(buffer, '\n');
		if(cp)
			*cp = 0;

		cp = strchr(buffer, ':');
		if(cp)
			*(cp++) = 0;

		if(cp && *cp)
			mode = cp;
		realm = strdup(buffer);
	}

	if(*argv)
		goto usage;

	digestbuf = (string_t)user + ":" + (string_t)realm + ":" + (string_t)secret;

	if(!stricmp(mode, "sha1"))
		digest::sha1(digestbuf);
	else if(!stricmp(mode, "rmd160"))
		digest::rmd160(digestbuf);
	else
		digest::md5(digestbuf);

	if(digestbuf[0] == 0)  {
		fprintf(stderr, "*** sipdigest: unsupported computation\n");
		exit(6);
	}

	printf("%s\n", *digestbuf);
	exit(0);
}

